pkg/logentry/stages/multiline_test.go

```diff
@@ -0,0 +1,40 @@
+package stages
+
+import (
+	"bytes"
+	"testing"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/prometheus/common/model"
+	"github.com/stretchr/testify/require"
+	ww "github.com/weaveworks/common/server"
+)
+func Test_multilineStage_Process(t *testing.T) {
+	// Enable debug logging
+	cfg := &ww.Config{}
+	require.Nil(t, cfg.LogLevel.Set("debug"))
+	util.InitLogger(cfg)
+	Debug = true
+
+	mcfg := &MultilineConfig{Expression: ptrFromString("^START"),}
+	validateMultilineConfig(mcfg)
+	stage := &multilineStage{
+		cfg: mcfg,
+		logger: util.Logger,
+		buffer: new(bytes.Buffer),
+	}
+
+	stage.Process(model.LabelSet{}, map[string]interface{}{}, ptrFromTime(time.Now()), ptrFromString("START line 1"))
+	stage.Process(model.LabelSet{}, map[string]interface{}{}, ptrFromTime(time.Now()), ptrFromString("not a start line"))
+
+	nextStart := "START line 2"
+	stage.Process(model.LabelSet{}, map[string]interface{}{}, ptrFromTime(time.Now()), &nextStart)
+
+	require.Equal(t, "START line 1\nnot a start line", nextStart)
```

Comment:
This is probably the confusing part. A new line is pushed but we get the older multiline blog. Ideally a pipeline stage would be able to push out lines without a call. This could be done with [channels](https://blog.golang.org/pipelines).

Reply:
Once we are processing inputs from channels and push changes out to channels the multiline stage could add  a simple timeout to flush a block:
```go

struct  Entry type {
    labels model.LabelSet,
    extracted map[string]interface{},
    t *time.Time,
    entry *string
}

// Process implements Stage
func (m *multilineStage) Process(entries <- chan Entry) <- chan Entry {
  out := make(chan Entry)
  currentMultiline := make(bytes.Buffer)
 
  loop { // This should be in a goroutine
    select {
    case next := <-entries:
        if isNewStart(entry) {
            // Flush
            out <- NewEntry(currentMultiline.String())
            currentMultiline.reset()
        }
        currentMultiline.WriteString(entry.entry)
    case <-time.After(timeout * time.Second):
            // Flush
            out <- currentMultiline.String()
            currentMultiline.reset()
    }
  }
  return out
}
```

`Process` is not a fitting name any more since it is called once for all incoming entries. Anyhow, I think it illustrates how the flush time out becomes much simpler.